probe-rs/src/flashing/builder.rs

```diff
@@ -0,0 +1,697 @@
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+pub(super) struct FlashPage {
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) address: u32,
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) size: u32,
+    #[derivative(Debug(format_with = "fmt"))]
+    pub(super) data: Vec<u8>,
+}
+
+fn fmt(data: &[u8], f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+    write!(f, "{:?}", data)
+}
+
+fn fmt_hex<T: std::fmt::LowerHex>(
+    data: &T,
+    f: &mut std::fmt::Formatter,
+) -> Result<(), std::fmt::Error> {
+    write!(f, "0x{:08x}", data)
+}
+
+impl FlashPage {
+    fn new(page_info: &PageInfo) -> Self {
+        Self {
+            address: page_info.base_address,
+            size: page_info.size,
+            data: vec![0; page_info.size as usize],
+        }
+    }
+}
+
+/// A struct to hold all the information about one Sector in the flash.
+#[derive(Derivative)]
+#[derivative(Debug, Clone)]
+pub(super) struct FlashSector {
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) address: u32,
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) size: u32,
+}
+
+impl FlashSector {
+    /// Creates a new empty flash sector form a `SectorInfo`.
+    fn new(sector_info: &SectorInfo) -> Self {
+        Self {
+            address: sector_info.base_address,
+            size: sector_info.size,
+        }
+    }
+}
+
+/// A struct to hold all the information about one region
+/// in the flash that is erased during flashing and has to be restored to it's original value afterwards.
+#[derive(Derivative)]
+#[derivative(Debug, Clone)]
+pub(super) struct FlashFill {
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) address: u32,
+    #[derivative(Debug(format_with = "fmt_hex"))]
+    pub(super) size: u32,
+    pub(super) page_index: usize,
+}
+
+impl FlashFill {
+    /// Creates a new empty flash fill.
+    fn new(address: u32, size: u32, page_index: usize) -> Self {
+        Self {
+            address,
+            size,
+            page_index,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct FlashLayout {
+    sectors: Vec<FlashSector>,
+    pages: Vec<FlashPage>,
+    fills: Vec<FlashFill>,
+}
+
+impl FlashLayout {
+    pub(super) fn sectors(&self) -> &[FlashSector] {
+        &self.sectors
+    }
+
+    pub(super) fn pages(&self) -> &[FlashPage] {
+        &self.pages
+    }
+
+    pub(super) fn pages_mut(&mut self) -> &mut [FlashPage] {
+        &mut self.pages
+    }
+
+    pub(super) fn fills(&self) -> &[FlashFill] {
+        &self.fills
+    }
+}
+
+#[derive(Clone, Copy)]
+pub(super) struct FlashDataBlock<'a> {
+    address: u32,
+    data: &'a [u8],
+}
+
+impl<'a> FlashDataBlock<'a> {
+    fn new(address: u32, data: &'a [u8]) -> Self {
+        Self { address, data }
+    }
+
+    pub(super) fn address(&self) -> u32 {
+        self.address
+    }
+
+    pub(super) fn size(&self) -> u32 {
```

Comment:
Hm I am not sure I want to change this, actually. Will have to think about it :) I chose u32 because that's the limit of the actual size. Maybe that was bad.
As for len: isn't that more reserved for general purpose structures and their actual length and not their memory size?